### ISSUE TITLE
How to delete/remove PagedListAdapter item

### ISSUE THREAD

**Issue by faris-jameel** [android/architecture-components-samples#281]
Currently I am using Android Architecture Components for App development everything is working along with paging library Now I want to remove recyclerview Item using PagedListAdapter to populate this we required to add a data source and from data source list is updating using LiveData no I want to remove a item from list notifyItemRemoved() is working from PagedList I am getting this exception:

java.lang.UnsupportedOperationException

java.util.AbstractList.remove(AbstractList.java:638)

**Comment by JiajunWong** [comment]
I had same issue.

**Comment by faris-jameel** [comment]
@JiajunWong did you find any solution?

**Comment by tuinui** [comment]
i got this problem to when i try to Collections.sort

**Comment by guness** [comment]
I have created some custom adapter to handle this case. Luckily I was supposed to add on top of current list, so following solution only supports adding to top. However you can try to customize it.

https://gist.github.com/guness/df12d8cc4f595af1395f4a1f5bca5f00

In addition, I think some kind of merging support for the data sources would be a nice to have feature for pagination library.

**Comment by JakeBase** [comment]
i had same issue

**Comment by shashankkapsime** [comment]
Can anyone answer this issue please.

**Comment by Tenkei7** [comment]
Already had this discussion [here](https://plus.google.com/communities/105153134372062985968/posts/108743107196401803953):

> If you have a collection shared w/ a RecyclerView, that collection CANNOT be modified w/o telling RecyclerView. So all modification & notifications needs to happen on the main thread and in the same call stack.
> 
> We could technically create a DataSource API that allows passing such partial changes but then we would need to log all of them and re-deliver to the RV on the main thread. The issue w/ that approach is that, if you have a RecyclerView that is stopped (a.k.a back stack), it will not (should not) receive any updates so PagedList would be keeping this potentially long list of items and re-apply on main thread for each observer (again some observers may have received some events).
> 
> This greatly complicates the problem, which is why we go w/ a single list.
> 
> Also, none of those optimizations would work w/ sqlite (so yea, that had a part). You can easily create a DataSource that can take a previous list and keep modifications on it, but still return it as a list. In other words,  implement that tracking logic in your DataSource which would not need to deal w/ RecyclerView's thread restrictions.

TL;DR: To update a pagedList you have to create a new list, this is how room update data and you have to do the same if you want to update your list.

**Comment by CarGuo** [comment]
I try this after delete local data
```
liveModel?.pageList?.dataSource?.invalidate()
```

**Comment by AdamSHurwitz** [comment]
After re-reading the documentation under [**Consider How Content Updates Work**](https://developer.android.com/topic/libraries/architecture/paging/data#consider-content-updates) it appears the only way to have realtime updates is via implementing Room with the PagedList: _If you're loading data directly from a Room database updates get pushed to your app's UI automatically._

**Comment by abhinav272** [comment]
@AdamSHurwitz 
If you connect your server data with `Room` to create `PagedList` from it, even then when you delete an item from your local DB your `DataSource` will be invalidated automatically and you will get another version of `PagedList` from Room in your onChanged() 
Am I right?

**Comment by AdamSHurwitz** [comment]
@abhinav272 This is not the case. When an individual item in **Room** is added or modified the **PagedList** updates the corresponding entry rather than refreshing all of the data in the component. Therefore I'm able to update Room with Firestore data when it changes and allow Room + PagedList to handle the updates.

Conversely, if Firestore data was connected directly to the PagedList the entire data set would need to be _invalidated_ / refreshed since the PagedList cannot automatically recognize individual changes from Firestore.

**Comment by abhinav272** [comment]
@AdamSHurwitz
Do you get a callback in onChanged() when you update the data in Room??
If so, that means Room has invalidated the DataSource and has provided new one with PagedList containing new data.

When I create a DataSource which is fetching data from either Room or Network, it gives me a callback in onChanged() only once when the PagedList gets created (upon creation of DataSource) and when the pages are loaded it automatically gets displayed on UI..

Make sense?

**Comment by AdamSHurwitz** [comment]
@abhinav272 - May you specify which **onChangedd()** method you are referring to?

I don't make Room updates directly. I update Firestore's database. With Firestore when a piece of information changes the Firestore listener will send an update. I listen for that update and inform Room of the change using a **LiveData** object. This makes it easy to keep both my backend Firestore and front-end Room data synced with minimal work.

**Comment by Dilip23** [comment]
@abhinav272 

  I had the same kind of issue!!!
I am building an app that receives data from network API which gets chached in Room DB. Now if i have to delete an item ,I send a request to delete an item in Remote DB(Network) and if the response is 204 , I delete an item in the Room , So then the Room DB creates a new PagedList and notifies the UI.... Right??

Or else Do we have to implement in any other way???

**Comment by abhinav272** [comment]
@Dilip23 yeah, this looks fine.

**Comment by Dilip23** [comment]
@abhinav272 

So this means I do not have to notify either Adapter of the removed item or invalidate the DataSource.

**Comment by roby222** [comment]
@faris-jameel if I understood correctly the question...
If you use the PagedListAdapter, you don't need to call any notifyItemRemoved() or similar.
So, if you delete your item from Room, the recycler adapter will be updated automatically

**Comment by abhinav272** [comment]
@sonalchopra-vvdn 
There may be a problem with your DiffUtil.ItemCallback<>
Fyi, whole list is returned Everytime whether you add/remove/update any item at any position, but the PagedListAdapter is smart enough to draw only those items which are new or changed, it skips re-drawing items which are same and this logic is implemented in your DiffUtil.ItemCallback<> overridden methods.

**Comment by roby222** [comment]
I made a sample using PagedListAdapter https://github.com/roby222/recyclerViewSample
I'm currently work on that because the page size settings doesn't work (I setted 20 items at a time, but I receive all db items!)

**Comment by alexandru-calinoiu** [comment]
I would love to see an example of this using anything else then room.

**Comment by parcool** [comment]
I need a sample that without `room` too.

**Comment by frog1014** [comment]
need a sample that without room too.

**Comment by mochadwi** [comment]
> I have created some custom adapter to handle this case(only a part of it actually). Luckily I was supposed to add on top of current list, so following solution only supports adding to top. However you can try to customize it.
> 
> https://gist.github.com/guness/df12d8cc4f595af1395f4a1f5bca5f00
> 
> In addition, I think some kind of merging support for the data sources would be a nice to have feature for pagination library.

@guness Maybe you want to update the class with `PagedListAdapterHelper`

**Comment by dlam** [comment]
We're discussing options to better support granular updates in Paging3, but currently to delete / remove an item you'll need to update the backing db, then invalidate your `DataSource` / `PagingSource`. Room handles propagating invalidation for you, but if you're not using Room you'll need to manually call `DataSource.invalidate()`. Your `DataSouce.Factory` should then generate a new instance of `DataSource` to load the updated data.

**Comment by dlam** [comment]
In Paging3, we also provide APIs to transform a `PagingData` (new `PagedList`) as we expose `PagingData` as a `Flow`. You can theoretically combine / filter / map / etc, in whatever way your heart desires, but the backing mechanism is still the same. We're looking into ways to provide more granular updates without invalidating the whole list as it's a common use case, but Paging3 is still very early in alpha.

**Comment by ishdemon** [comment]
so there is just no way to update/modify a pagedlist item without invalidating the datasource if i am not using room. -_-

**Comment by dlam** [comment]
Actually even with Room it works the exact same way, Room just handles the invalidation for you. Any modifications you make also need to update the source of truth for paging, because reloading the page should always keep the updates you make.

For a `Flow` / stream based approach which allows Paging to subscribe to individual item / page updates that do not require the invalidation loop please follow / +1 this bug: https://issuetracker.google.com/160232968

**Comment by 309152665** [comment]
I need a sample that without room too.

**Comment by SkyairOnline** [comment]
Well if you want to do that, maybe you can make the visibility into gone if you want to delete, it will:
1. Delete item in background real time
2. Make the viewer believe that item has already been deleted

Ex: itemView.visibility = View.gone();

But It actually is not a best practice, so until the efficient solution is presented, I use that, But I actually recommended to use local db for best practice

**Comment by ahmednabeel1991** [comment]
Hello All,

After Allot of search I fixed this bug. The answer is here bellow

    mainListAdapter.snapshot().toMutableList().apply { removeAt(position) }
                mainListAdapter.notifyItemRemoved(position)
                
                Cheers!!

**Comment by guness** [comment]
This answer assumes that the list used in `mainListAdapter` is mutable. However, it does not have to be one. So the solution may start failing on another version.

**Comment by dlam** [comment]
> Hello All,
> 
> After Allot of search I fixed this bug. The answer is here bellow
> 
>     mainListAdapter.snapshot().toMutableList().apply { removeAt(position) }
>                 mainListAdapter.notifyItemRemoved(position)
>                 
>                 Cheers!!

This won't work, you must remove items through Paging via invalidation, otherwise Paging will not know to carry your otem removal through page reloads, refresh, config changes, etc.

**Comment by 309152665** [comment]
> Hello All,
> 
> After Allot of search I fixed this bug. The answer is here bellow
> 
> ```
> mainListAdapter.snapshot().toMutableList().apply { removeAt(position) }
>             mainListAdapter.notifyItemRemoved(position)
>             
>             Cheers!!
> ```

thanks , but the network data with pagingdata , snapshot is the pagingdata's copy.   snapshot.toMutableList() , the mutablelist is the  snapshot copy.

**Comment by 309152665** [comment]
> This answer assumes that the list used in `mainListAdapter` is mutable. However, it does not have to be one. So the solution may start failing on another version.

ok , thanks

**Comment by ahmednabeel1991** [comment]
> > Hello All,
> > After Allot of search I fixed this bug. The answer is here bellow
> > ```
> > mainListAdapter.snapshot().toMutableList().apply { removeAt(position) }
> >             mainListAdapter.notifyItemRemoved(position)
> >             
> >             Cheers!!
> > ```
> 
> thanks , but the network data with pagingdata , snapshot is the pagingdata's copy. snapshot.toMutableList() , the mutablelist is the snapshot copy.

If you are using only network call with pagging then this solution will work because you will remove Item locally after network api call . Let say I Call retrofit call first for removing the item and on success response of that Api Call we will remove that particular item from snapshot.

**Comment by 309152665** [comment]
> > > Hello All,
> > > After Allot of search I fixed this bug. The answer is here bellow
> > > ```
> > > mainListAdapter.snapshot().toMutableList().apply { removeAt(position) }
> > >             mainListAdapter.notifyItemRemoved(position)
> > >             
> > >             Cheers!!
> > > ```
> > 
> > 
> > thanks , but the network data with pagingdata , snapshot is the pagingdata's copy. snapshot.toMutableList() , the mutablelist is the snapshot copy.
> 
> If you are using only network call with pagging then this solution will work because you will remove Item locally after network api call . Let say I Call retrofit call first for removing the item and on success response of that Api Call we will remove that particular item from snapshot.

snapshot is copy , delete from snapshot , the pagingData not change.

**Comment by ahmednabeel1991** [comment]
> > > > Hello All,
> > > > After Allot of search I fixed this bug. The answer is here bellow
> > > > ```
> > > > mainListAdapter.snapshot().toMutableList().apply { removeAt(position) }
> > > >             mainListAdapter.notifyItemRemoved(position)
> > > >             
> > > >             Cheers!!
> > > > ```
> > > 
> > > 
> > > thanks , but the network data with pagingdata , snapshot is the pagingdata's copy. snapshot.toMutableList() , the mutablelist is the snapshot copy.
> > 
> > 
> > If you are using only network call with pagging then this solution will work because you will remove Item locally after network api call . Let say I Call retrofit call first for removing the item and on success response of that Api Call we will remove that particular item from snapshot.
> 
> snapshot is copy , delete from snapshot , the pagingData not change.

Try this with the same code it is working.

**Comment by 309152665** [comment]
> > > > > Hello All,
> > > > > After Allot of search I fixed this bug. The answer is here bellow
> > > > > ```
> > > > > mainListAdapter.snapshot().toMutableList().apply { removeAt(position) }
> > > > >             mainListAdapter.notifyItemRemoved(position)
> > > > >             
> > > > >             Cheers!!
> > > > > ```
> > > > 
> > > > 
> > > > thanks , but the network data with pagingdata , snapshot is the pagingdata's copy. snapshot.toMutableList() , the mutablelist is the snapshot copy.
> > > 
> > > 
> > > If you are using only network call with pagging then this solution will work because you will remove Item locally after network api call . Let say I Call retrofit call first for removing the item and on success response of that Api Call we will remove that particular item from snapshot.
> > 
> > 
> > snapshot is copy , delete from snapshot , the pagingData not change.
> 
> Try this with the same code it is working.

the code is working , but you can try more items adapter , 
![device-2021-07-14-174201](https://user-images.githubusercontent.com/7857673/125600785-74b8fc3f-99d9-41b0-8700-0940cca515df.png)
such as this record , over a screen items , you delete the first one , scroll and look , the items in last , you can find some items has disappear.
![image](https://user-images.githubusercontent.com/7857673/125601510-fc92001a-72b7-4d33-8db6-99bf49f6a39e.png)

https://user-images.githubusercontent.com/7857673/125601652-fa882d61-c084-4d39-8dfe-1da102ece895.mp4

**Comment by rajangupt1008** [comment]
I have implemented PagingDataAdapter in my project and required multiple filtering and sorting. so any one could suggest me achive this.

**Comment by Reypak** [comment]
> .toMutableList().apply

please help convert to java

**Comment by dharamveer-appyhigh** [comment]
> I try this after delete local data
> 
> ```
> liveModel?.pageList?.dataSource?.invalidate()
> ```

But what if we are using cachedIn Flow
